### PR TITLE
Subpackages concepts and guides

### PR DESCRIPTION
**Pull request by phanimarupaka**
This section describes the concept around nested packages and adds guides.

### CODE REVIEW

File: site/content/en/concepts/packaging/_index.md
Comment (mortent) [inline site/content/en/concepts/packaging/_index.md:36]: Instead of composite package here, I think we can mark this one as `top-level package` and the others as `nested package`?

File: site/content/en/concepts/packaging/_index.md
Comment (mortent) [inline site/content/en/concepts/packaging/_index.md:40]: At least point out that this is also a package.

File: site/content/en/concepts/packaging/_index.md
Comment (mortent) [inline site/content/en/concepts/packaging/_index.md:57]: This is not something we have implemented yet. Adding this to the docs now would be confusing.

File: site/content/en/concepts/packaging/_index.md
Comment (mortent) [inline site/content/en/concepts/packaging/_index.md:54]: I'm not sure if what we are providing today fits this description. We provide a way to create setters across nested packages, but we don't really do any enforcement.

**Review by mortent — COMMENTED** [review]
So I think the main question here is whether we want to call this composite packages or nested packages. I think the latter better describes our current functionality.

File: site/content/en/concepts/packaging/_index.md
Comment (mortent) [inline site/content/en/concepts/packaging/_index.md:36]: So I'm thinking thinking that this is not a nested package, but rather the top-level package. I'm thinking nested package and subpackage means the same.

**Review by mortent — COMMENTED** [review]
One question about terminology, but overall looks pretty good.

File: site/content/en/concepts/packaging/_index.md
Comment (frankfarzan) [inline site/content/en/concepts/packaging/_index.md:67]: Not sure I follow #2. What do you mean by actions? Kpt functions act on resources of subpackage for example.

File: site/content/en/concepts/packaging/_index.md
Comment (frankfarzan) [reply site/content/en/concepts/packaging/_index.md:36]: There are two dimensions.

1. A package may or may not contain other packages.
2. A subpackage may be declared using `Kptfile.dependencies` or by virtue of containing a subpackage dir.

For #1, I think we just need to define the term 'subpackage' and avoid 'composite' or 'nested', etc.

Not sure what we want to call #2, do we have terms for them? `dependencies` is going to be overloaded with eventual plans around declarative Pipelines.

Comment (phanimarupaka) [reply site/content/en/concepts/packaging/_index.md]: Removed all the terms `composite` or `nested`, etc. Renamed the concept to `subpackages` and explained it. Guides are also updated accordingly. `Kptfile.dependencies` is not documented officially yet, so restricted the discussion to packages in the directory tree of another package. We can expand more once we ship `pkg` features. PTAL.

File: site/content/en/concepts/packaging/_index.md
Comment (phanimarupaka) [reply site/content/en/concepts/packaging/_index.md:67]: Added that this is only supported for `cfg` commands currently.

File: site/content/en/concepts/packaging/_index.md
Comment (phanimarupaka) [reply site/content/en/concepts/packaging/_index.md]: Removed.

**Review by mortent — APPROVED** [review]

File: package-examples/subpackages/wordpress/README.md
Comment (frankfarzan) [inline package-examples/subpackages/wordpress/README.md:67]: sorry for being a stickler here: Are we just referring to 'setter parameters' everywhere? I think you want to reduce as many redundant words and concepts as possible. So:

- In one place (in setters documentation) state that setters are how kpt implements parameterization.
- In rest of kpt docs, just refer to it as 'setters'.

Phrasing to be used consistency everywhere:

A producer "declares a setter definition"
A consumer "provides a setter value". Currently this is imperative, but in the future it will be declarative: "Declare setter values".

Make the changes as a follow up. Don't block this PR though.

File: site/content/en/concepts/packaging/_index.md
Comment (frankfarzan) [inline site/content/en/concepts/packaging/_index.md:67]: nit: `package (e.g. `bar`)`

Space before `(`
Use `e.g.`

Consistent everywhere.

File: site/content/en/concepts/packaging/_index.md
Comment (frankfarzan) [inline site/content/en/concepts/packaging/_index.md:70]: Just say to run a command recursively. "To perform some action" is unnecessarily vague.

Comment (frankfarzan) [inline site/content/en/concepts/packaging/_index.md:70]: This is not a "principle" per se, but just a CLI design choice.

File: site/content/en/concepts/packaging/_index.md
Comment (frankfarzan) [inline site/content/en/concepts/packaging/_index.md:67]: Is #2 and #3 just saying this:

By default, commands performed on a package are not performed on subpackages unless -R is provided

?

If so,  remove unnecessary fluff.

File: site/content/en/guides/consumer/subpackages/_index.md
Comment (frankfarzan) [inline site/content/en/guides/consumer/subpackages/_index.md:43]: I have comments about the structure of this in the friction log.

File: package-examples/subpackages/wordpress/README.md
Comment (phanimarupaka) [reply package-examples/subpackages/wordpress/README.md]: Agree. Simplified all the text and added references rather than redefining the concepts. Thanks.

File: site/content/en/guides/consumer/subpackages/_index.md
Comment (phanimarupaka) [reply site/content/en/guides/consumer/subpackages/_index.md]: Addressed them. PTAL.

File: site/content/en/concepts/packaging/_index.md
Comment (phanimarupaka) [reply site/content/en/concepts/packaging/_index.md]: Agree. Consolidated 2 and 3.

File: package-examples/subpackages/wordpress/README.md
Comment (frankfarzan) [inline package-examples/subpackages/wordpress/README.md:60]: like discussed: "Provide values for all required setters"

File: package-examples/subpackages/wordpress/README.md
Comment (frankfarzan) [inline package-examples/subpackages/wordpress/README.md:63]: Rephrase:

"By default, the 'set' command is performed
  only on the resource files of parent package and not the subpackages. This behavior may be changed by setting  `--recurse-subpackages(-R)`..."

General comments
- First describe the behavior. How that behavior is changed is a detail that follows.
- Be consistent between 'operation', 'action', and 'command'. It's used interchangeably and is confusing. In this context, CLIs have 'commands'.

Apply these consistency everywhere.

**Review by frankfarzan — APPROVED** [review]
lgtm after comments are addressed

**Pull request merged**
